src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -100,14 +100,39 @@ public bool HasCompilation
                 }
             }
 
-            public bool ContainsAssemblyOrModuleOrDynamic(ISymbol symbol)
+            /// <summary>
+            /// Returns <see langword="true"/> if this <see cref="Project"/>/<see cref="Compilation"/> could produce the
+            /// given <paramref name="symbol"/>.  The symbol must be a <see cref="IAssemblySymbol"/>, <see
+            /// cref="IModuleSymbol"/> or <see cref="IDynamicTypeSymbol"/>.
+            /// </summary>
+            /// <remarks>
+            /// If <paramref name="primary"/> is true, then <see cref="Compilation.References"/> will not be considered
+            /// when answering this question.  In other words, if <paramref name="symbol"/>  is an <see
+            /// cref="IAssemblySymbol"/> and <paramref name="primary"/> is <see langword="true"/> then this will only
+            /// return true if the symbol is <see cref="Compilation.Assembly"/>.  If <paramref name="primary"/> is
+            /// false, then it can return true if <paramref name="symbol"/> is <see cref="Compilation.Assembly"/> or any
+            /// of the symbols returned by <see cref="Compilation.GetAssemblyOrModuleSymbol(MetadataReference)"/> for
+            /// any of the references of the <see cref="Compilation.References"/>.
+            /// </remarks>
+            public bool ContainsAssemblyOrModuleOrDynamic(ISymbol symbol, bool primary)
             {
                 Debug.Assert(symbol.Kind == SymbolKind.Assembly ||
                              symbol.Kind == SymbolKind.NetModule ||
                              symbol.Kind == SymbolKind.DynamicType);
                 var state = this.ReadState();
                 var unrootedSymbolSet = state.UnrootedSymbolSet;
-                return unrootedSymbolSet?.Contains(symbol) ?? false;
+                if (unrootedSymbolSet == null)
+                    return false;
+
+                if (primary)
+                {
+                    return symbol.Equals(unrootedSymbolSet.Value.PrimaryAssemblySymbol.GetTarget()) ||
+                           symbol.Equals(unrootedSymbolSet.Value.PrimaryDynamicSymbol.GetTarget());
+                }
+                else
+                {
+                    return unrootedSymbolSet.Value.SecondaryReferencedSymbols.Contains(symbol);
+                }
```

Comment:
One core issue was that when we looking for the right project to use for Assembly 'Foo', we might return project 'Bar' if Bar had a compilation-reference to Foo.  This is not a problem for *top-level* symbols, since we'll be able to find that top level symbol in either project, however it was a major problem for body level symbols as those symbols can only be found in 'Foo' and not in 'Bar'.  Indeed, body-level symbols attempt to use their syntax-tree info to find the symbol again.  Those trees are only valid in their originating project, so we must map those back to their originating project, and not a project that references that project.